### PR TITLE
PR#5

### PR DESCRIPTION
**Pull request by jimmyhurricanedang**
- Added feature to add a location to a route
- its possible to give a location a description
- fixed a few bugs

### CODE REVIEW

File: frontend/src/pages/RouteDetails.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+export default function RouteDetails (){
+    const classes = useStyles();
+
+    const { name } = useParams()
+    const [routeData, setRouteData] = useState()
+
+
+    useEffect(() => {
+        getRoute(name)
+            .then(setRouteData)
```
Comment (jamarob) [inline frontend/src/pages/RouteDetails.js:33]: Error handling?

File: frontend/src/pages/RouteDetails.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+    return (
+        <RouteDetailsContainer>
+            <Card className={classes.root} variant="outlined">
+                <CardContent>
+                    <Typography className={classes.title} color="textSecondary" gutterBottom>
+                        Creator: {routeData.creatorUserName}
+                    </Typography>
+                    <Typography variant="h5" component="h2"> Routename: {routeData.name}</Typography>
+                    <Typography className={classes.pos} color="textSecondary"> Country: {routeData.country} </Typography>
+                    <Typography variant="body2" component="p">Visited Locations: {routeData?.locations?.map(location => <p>{location.locationName}</p>)}</Typography>
```
Comment (jamarob) [inline frontend/src/pages/RouteDetails.js:53]: route data will be defined at this point, not need for the optional chaining

File: frontend/src/pages/OverviewPage.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+            <button>
+                <Link to="/routesoverview">
+                    Show all Routes
+                </Link>
+            </button>
+            <button>
+                <Link to="/newroutepage">
+                    Add new Route
+                </Link>
+            </button>
```
Comment (jamarob) [inline frontend/src/pages/OverviewPage.js:17]: No links in buttons and vice versa, you can always style your links to look like buttons.

File: frontend/src/pages/OverviewPage.js [hunk 4]
```diff
@@ -0,0 +1,20 @@
+
+import {Link} from "react-router-dom";
+import React from "react";
+
+export default function OverviewPage() {
+    return (
+        <div>
```
Comment (jamarob) [inline frontend/src/pages/OverviewPage.js:7]: Maybe use a `nav` here?

File: frontend/src/pages/NewRoutePage.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
+import AddNewRouteForm from "../components/AddNewRouteForm";
+import {postRoute} from "../service/columbusApiService";
+import {useState} from "react";
+import MapContainer from "../components/MapContainer";
+
+const initState = {name: "", country: "", creatorUserName: "", locations: [] }
```
Comment (jamarob) [inline frontend/src/pages/NewRoutePage.js:6]: 👍

File: frontend/src/pages/NewRoutePage.js [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+export default function NewRoutePage() {
+    const [routes, setRoutes] = useState([])
+    const [routeToAdd, setRouteToAdd] = useState(initState)
+
+    // set markers onClick on the map
```
Comment (jamarob) [inline frontend/src/pages/NewRoutePage.js:12]: please delete

File: frontend/src/pages/NewRoutePage.js [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+            .catch((error) => console.error(error))
+    }
+
+
+
+    return (
+        <>
+            <MapContainer markers={markers} setMarkers={setMarkers} className="mapContainer"/>
+            <AddNewRouteForm onSubmit={handleSubmit} routeToAdd={routeToAdd} handleChange={handleChange}/>
+            <div>{markers.map(marker => <><p>{marker.lat}</p><p>{marker.locationName}</p></>)}</div>
```
Comment (jamarob) [inline frontend/src/pages/NewRoutePage.js:36]: use `React.Fragment` instead of `<>` so you can assign a key prop

File: frontend/src/index.css [hunk 8]
```diff
@@ -21,3 +23,20 @@ code {
...
+  max-width: 400px;
+  z-index: 10;
+}
+
+.search input {
+  padding: 0.5rem;
+  font-size: 1.5rem;
+  width: 100%;
+  text-align: center;
+}
```
Comment (jamarob) [inline frontend/src/index.css:42]: I think this belongs inside a component, not a global css file. Think about using styled-components and putting it in `GlobalStyle`

File: frontend/src/components/RouteListItem.js [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+
+    return (
+            <Card className={classes.root} variant="outlined">
+                <CardContent>
+                    <Typography className={classes.title} color="textSecondary" gutterBottom>
+                        Creator: {route.creatorUserName}
+                    </Typography>
+                    <Typography variant="h5" component="h2"> Routename: {route.name}</Typography>
+                    <Typography className={classes.pos} color="textSecondary"> Country: {route.country} </Typography>
+                    <Typography variant="body2" component="p">Visited Locations: {route?.locations?.map(location => <p>{location.locationName}</p>)}</Typography>
```
Comment (jamarob) [inline frontend/src/components/RouteListItem.js:33]: why use optional chaining here and not before? Please ad a key prop when you map

File: frontend/src/components/MapContainer.js [hunk 10]
```diff
@@ -0,0 +1,134 @@
+import React from 'react';
+import {GoogleMap, useLoadScript, InfoWindow} from '@react-google-maps/api';
+import {useState, useCallback, useRef} from 'react'
+import PanToCurrentLocation from "./Map/PanToCurrentLocation";
+import {formatRelative} from 'date-fns';
+import RouteConnector from "./Map/RouteConnector";
+import Search from "./Map/Search";
+
+
+// additional google libraries; "places" for the search function on the map
```
Comment (jamarob) [inline frontend/src/components/MapContainer.js:10]: please delete

File: frontend/src/components/MapContainer.js [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+const libraries = ['places']
+
+// set size of the rendered map
```
Comment (jamarob) [inline frontend/src/components/MapContainer.js:13]: please delete

File: frontend/src/components/MapContainer.js [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+const mapContainerStyle = {
+    width: "100vw",
+    height: "50vh",
+};
+
+//Default Location of Oldenburg
```
Comment (jamarob) [inline frontend/src/components/MapContainer.js:19]: please delete

File: frontend/src/components/MapContainer.js [hunk 13]
```diff
@@ -0,0 +1,134 @@
...
+    disableDefaultUI: true,
+    mapTypeControl: true,
+    zoomControl: true,
+};
+
+
+
+export default function MapContainer({markers, setMarkers}) {
+
+    // script to load the map + libraries
```
Comment (jamarob) [inline frontend/src/components/MapContainer.js:36]: please delete

File: frontend/src/components/MapContainer.js [hunk 14]
```diff
@@ -0,0 +1,134 @@
...
+    const {isLoaded, loadError} = useLoadScript({
+        googleMapsApiKey: process.env.REACT_APP_GOOGLE_MAPS_API_KEY,
+        version: '3.42.9',
+        libraries,
+    });
+
+
+    // infoWindow for selected marker
```
Comment (jamarob) [inline frontend/src/components/MapContainer.js:44]: please delete

File: frontend/src/components/MapContainer.js [hunk 15]
```diff
@@ -0,0 +1,134 @@
...
+    const [selected, setSelected] = useState(null);
+
+
+
+    // prevent map to trigger a re-render
```
Comment (jamarob) [inline frontend/src/components/MapContainer.js:49]: please delete

File: frontend/src/components/MapContainer.js [hunk 16]
```diff
@@ -0,0 +1,134 @@
...
+        setMarkers((current) => current.map(marker => {
+            if (marker.lat === selected.lat && marker.lng === selected.lng){
+                return {...marker, locationName: event.target.value,}
+            }
+            return marker;
+        }))
+    }
+
+
+    // makes map re-center to new position and prevents re-render
```
Comment (jamarob) [inline frontend/src/components/MapContainer.js:73]: please delete

File: frontend/src/components/Map/Search.js [hunk 17]
```diff
@@ -0,0 +1,57 @@
+import usePlacesAutocomplete, {getGeocode, getLatLng} from "use-places-autocomplete";
+import {Combobox, ComboboxInput, ComboboxList, ComboboxOption, ComboboxPopover} from "@reach/combobox";
+import "@reach/combobox/styles.css";
+import React from "react";
+
+
+export default function Search({panTo}) {
+    const {ready, value, suggestions: {status, data}, setValue, clearSuggestions,} = usePlacesAutocomplete({
+        requestOptions: {
+            // search center point
```
Comment (jamarob) [inline frontend/src/components/Map/Search.js:10]: please delete

File: frontend/src/App.js [hunk 18]
```diff
@@ -1,15 +1,35 @@
...
-      <Map/>
-      <Overview/>
-    </div>
+    <Router>
+        <Switch>
+            <Route exact path="/">
+                <AppHeader/>
+                <div>
+                    <OverviewPage/>
+                </div>
```
Comment (jamarob) [inline frontend/src/App.js:20]: Why the `div`? If it is necessary put it into `OverviewPage`

File: frontend/src/App.js [hunk 19]
```diff
@@ -1,15 +1,35 @@
...
+            </Route>
+            <Route path="/newroutepage">
+                <NewRoutePage/>
+            </Route>
```
Comment (jamarob) [inline frontend/src/App.js:24]:
```suggestion
            <Route path="/newroutepage" component={NewRoutePage}/>
```
What do you think about this? short and sweet :)

File: frontend/src/App.css [hunk 20]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (jamarob) [inline frontend/src/App.css:1]: Please delete the css files and use styled-components

File: frontend/package.json [hunk 21]
```diff
@@ -43,6 +47,8 @@
     ]
   },
   "devDependencies": {
+    "@material-ui/icons": "^4.11.2",
```
Comment (jamarob) [inline frontend/package.json:50]: I think this is not a devDependency but needed for production

File: backend/src/test/java/de/jimmy/columbus/service/RouteServiceTest.java [hunk 22]
```diff
@@ -56,6 +55,25 @@ public void listRoutes() {
...
+                .build();
+
+        //WHEN
+        Optional<Route> routeByRouteName = routeService.getRouteByRouteName(actualRoute.getName());
+
+        //THEN
+        assertThat(routeByRouteName.get(), is(actualRoute));
+    }
+*/
+
```
Comment (jamarob) [inline backend/src/test/java/de/jimmy/columbus/service/RouteServiceTest.java:76]: Try to not pollute main with commented out code

File: backend/src/main/java/de/jimmy/columbus/model/Route.java [hunk 23]
```diff
@@ -20,11 +18,9 @@
 @Document(collection = "routes")
 public class Route {
     @Id
+    private String id;
```
Comment (jamarob) [inline backend/src/main/java/de/jimmy/columbus/model/Route.java:21]: `@Id` is not necessary when you have a field called `id`

**Review by jamarob — APPROVED** [review]
Good job! 🗺️ Easy to read and review. Left some comments, mostly on deleting comments ;)

File: frontend/src/pages/OverviewPage.js (hunk 3 above)
Comment (JimmyDang95) [reply frontend/src/pages/OverviewPage.js:17]: Ok, gonna keep that in mind ✌🏻

File: frontend/src/pages/NewRoutePage.js (hunk 7 above)
Comment (JimmyDang95) [reply frontend/src/pages/NewRoutePage.js:36]: Do I have to create that somewhere ? or can I simply replace the `<>` with `<React.Fragment>` ?

File: frontend/src/App.js (hunk 19 above)
Comment (JimmyDang95) [reply frontend/src/App.js:24]: Oh yeah ! looks much better !

**Pull request merged**